lib/ts/recipe/oauth2/types.ts

```diff
@@ -0,0 +1,298 @@
+              refreshToken?: string;
+              idToken?: string;
+              expiresIn: number;
+          }
+        | { status: "CLIENT_ERROR"; error: TokensErrorCode } // This results in a 400 with the error code added. The status specifies that this is a client error to be consistent with the createAuthCode
+    >;
+
+    verifyAccessToken(input: {
+        clientId: string;
+        accessToken: string;
+    }): Promise<{
+        sessionHandle?: string;
+        scopes: string[];
+        timeCreated: number;
+        lastUpdated: number;
+        timeAccessTokenExpires: number;
+        timeRefreshTokenExpires?: number;
+    }>;
+
+    getTokenInfo(input: {
+        clientId: string;
+        refreshToken?: string;
+        authCode?: string;
+    }): Promise<{
+        sessionHandle?: string;
+        scopes: string[];
+        timeCreated: number;
+        lastUpdated: number;
+        timeAccessTokenExpires: number;
+        timeRefreshTokenExpires?: number;
+    }>;
+
+    revokeToken(input: { clientId: string } | { sessionHandle: string } | { accessToken: string }): Promise<void>;
+    revokeAuthCodes(input: { clientId: string } | { sessionHandle: string }): Promise<void>;
+
+    // Customization points called by createTokens&createAuthCode
+    buildAccessToken(input: {
+        clientId: string;
+        userId?: string;
+        sessionHandle?: string;
+        scopes: string[];
+        userContext: any;
+    }): Promise<{ payload?: JSONObject; lifetimeInSecs?: number }>; // Returning undefined for these props means we use the default from Core
+    buildIdToken(input: {
+        clientId: string;
+        userId?: string;
+        sessionHandle?: string;
+        scopes: string[];
+        userContext: any;
+    }): Promise<{ payload?: JSONObject; lifetimeInSecs?: number }>; // Returning undefined for these props means we use the defaults from BE SDKs
+    getRefreshTokenLifetime(): Promise<number | undefined>; // Returning undefined means we use the default from Core
+
+    validateScopes(input: {
+        clientId: string;
+        userId?: string;
+        sessionHandle?: string;
+        scopes: string[];
+        userContext: any;
+    }): Promise<
+        | { status: "OK"; grantedScopes: string[] }
+        | { status: "INVALID_SCOPE_ERROR"; message: string } // Returning this signifies a client error and should result in a redirection back to the client with the "invalid_scope" error
+        | { status: "ACCESS_DENIED_ERROR"; message: string } // This means we need to redirect to the auth frontend (i.e.: there is a claim validation issue we can resolve there)
+    >;
+
+    shouldIssueRefreshToken(input: {
+        clientId: string;
+        userId?: string;
+        sessionHandle?: string;
+        scopes: string[];
+        userContext: any;
+    }): Promise<boolean>;
+};
+
+export type ResponseType = "code" | "id_token" | "token"; // We do not support "token", but it's a possible input value.
```

Comment:
if we don't supoprt token, we should not have it - otherwise what will the recipe function do if that is passed?

Reply:
It's important to note that it could be there (at least on the API level), since it's part of the protocol. We return with `unsupported_response_type`